stablehlo/tests/CheckOps.cpp

```diff
@@ -156,6 +157,96 @@ llvm::Error evalExpectSerializedEqOp(const Tensor &expected, StringRef probeId,
   return evalExpectEqOp(expected, *tensor);
 }
 
+static uint64_t ULPDifference(APFloat f, APFloat g) {
+  // z is ULP-distance between exact 0 and largest subnormal:
+  uint64_t z = APFloat::getSmallestNormalized(f.getSemantics())
+                   .bitcastToAPInt()
+                   .getLimitedValue() -
+               1;
+  // a is ULP-distance between exact 0 and abs(f):
+  uint64_t a = (f.isNegative() ? -f : f).bitcastToAPInt().getLimitedValue();
+  // b is ULP-distance between exact 0 and abs(g):
+  uint64_t b = (g.isNegative() ? -g : g).bitcastToAPInt().getLimitedValue();
+  if (f.isFinite() && g.isFinite()) {
+    // subtract subnormals contribution, round subnormals to closest normal or
+    // zero:
+    if (z < a)
+      a -= z;
+    else
+      a = (2 * a < z ? 0 : 1);
+    if (z < b)
+      b -= z;
+    else
+      b = (2 * b < z ? 0 : 1);
+    if (f.isNegative() != g.isNegative()) {
+      // in case of overflow, return uint64 maximal value
+      if (a > std::numeric_limits<uint64_t>::max() - b)
+        return std::numeric_limits<uint64_t>::max();
+      return a + b;
+    }
+  }
+  return (a > b ? a - b : b - a);
+}
+
+static uint64_t ULPDifference(const Element &e1, const Element &e2) {
+  // caller is responsible for ensuring that e1, e2 have both the same
+  // float or complex types
+  if (isSupportedComplexType(e1.getType())) {
+    auto complexLhs = e1.getComplexValue();
+    auto complexRhs = e2.getComplexValue();
+    return std::max(ULPDifference(complexLhs.real(), complexRhs.real()),
+                    ULPDifference(complexLhs.imag(), complexRhs.imag()));
+  } else {
+    return ULPDifference(e1.getFloatValue(), e2.getFloatValue());
+  }
+}
+
+llvm::Error evalExpectIsCloseToReferenceOp(const Tensor &actual,
+                                           const Tensor &reference,
+                                           const Tensor &input,
+                                           uint64_t max_ulp_difference) {
+  auto type = actual.getElementType();
+  if (!(isSupportedFloatType(type) || isSupportedComplexType(type)))
```

Comment:
I didn't notice the extra paren after the `!` at first, may be easier to read/reason about as: `!isSupportedFloatType(type) && !isSupportedComplexType(type)`

Reply:
Done.